src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializer.java

```diff
@@ -267,8 +267,10 @@ private final Object vanillaDeserialize(JsonParser p,
         final Object bean = _valueInstantiator.createUsingDefault(ctxt);
         // [databind#631]: Assign current value, to be accessible by custom serializers
         p.setCurrentValue(bean);
+        Set<String> presentNames = new HashSet<String>();
```

Comment:
Creating a new `HashSet` for each and every POJO ever deserialized adds serious overhead.
Related also to the general question of combining handling with `required` checking, since both could/should use a bitset of handling whether value for matching expected/defaulting property has been encountered.